### PR TITLE
Add exporting only pattern as PDF

### PR DESCRIPTION
**Pull request by cvick32**
Hi, 

  I opened issue #1756 because I wanted to print only the pattern into a PDF without any of the extra information that is normally provided by "Export PDF". 

  These changes add a new option "PDF Export Pattern Only" to the "Visualize and Export" menu. 

  Any feedback is welcome. In particular, I think that the new PrintPatternOnly class could inherit from Print and just set a class-wide "template" variable in the constructor to have the same effect. I might try this if I have some time, but I was just trying to get this working as fast as possible for my use case. 

 Thanks, Cole

### CODE REVIEW

**Comment by kaalleen** [comment]
Wow, nice, thank you for your work.

I think we could add this as a additional print layout page (such as operator overview, etc.) which can be switched on and off by the user in the print settings. Then we don't need a whole new extension. What do you think?

**Comment by cvick32** [comment]
Hey @kaalleen, I deleted the extra extension for the pattern-only view. Now, for some reason I can't get the pattern-only view to show/hide the view when the checkbox is checked. I can't really figure out the jquery and it looks like I'm doing everything that the other views are doing. 

Some help here would be much appreciated :).

**Comment by kaalleen** [comment]
You are very close to bring it to work.

* Remove `templates/print_pattern_only.xml`
* Remove your pattern-onlyview click function in inkstitch.js
* Open lib/extensions/print-pdf go to line 277 and add `'pattern_only': False,`

**Comment by cvick32** [comment]
So, I made those changes, but the "Pattern Only View" checkbox still isn't responsive on click like the other ones.

**Comment by kaalleen** [comment]
Oh right, I've been too fast. There is more, but as a start. Change in inkstitch.js from val() to checked: `settings["pattern-onlyview"] = $("[data-field-name='pattern-onlyview']").is(':checked');`

There is also updates needed in the index file, include like this:     <div class="page pattern-onlyview" style="display: {{ 'block' if view.pattern_onlyview else 'none' }}">{% include 'pattern_only.html' %}</div>

Then the data-field in pattern only needs to use -transform, in ui the minus between only-view should be gone, etc.

I'll answer more detailed tomorrow. I'm running out of time. But maybe also you can start with that and then work through it.

File: print/resources/inkstitch.js [hunk 1]
```diff
@@ -735,6 +735,7 @@ $(function() {
     settings["operator-overview"] = $("[data-field-name='operator-overview']").is(':checked');
     settings["operator-detailedview"] = $("[data-field-name='operator-detailedview']").is(':checked');
     settings["operator-detailedview-thumbnail-size"] = $("[data-field-name='operator-detailedview-thumbnail-size']").val();
+    settings["pattern-onlyview"] = $("[data-field-name='pattern-onlyview']").val();
```
Comment (kaalleen) [inline print/resources/inkstitch.js:738]: Change this to `settings["pattern-onlyview"] = $("[data-field-name='pattern-onlyview']").is(':checked');`

File: print/templates/index.html [hunk 2]
```diff
@@ -31,6 +31,9 @@
 {# operator detailed view #}
     {% include 'operator_detailedview.html' %}
 
+{# pattern only view #}
+    {% include 'pattern_only.html' %}
+
```
Comment (kaalleen) [inline print/templates/index.html:36]: Only include if this should actually show:

`<div class="page pattern-onlyview" style="display: {{ 'block' if view.pattern_onlyview else 'none' }}">{% include 'pattern_only.html' %}</div>`

File: print/resources/style.css [hunk 3]
```diff
@@ -492,6 +492,10 @@ body {
     padding-top: 2mm;
   }
 
+  .pattern-onlyview figure.inksimulation {
+      height: 100%;
+  }
+
```
Comment (kaalleen) [inline print/resources/style.css:498]: Remove this, we do not really need this twice

File: print/resources/style.css [hunk 4]
```diff
@@ -573,6 +577,11 @@ body {
     overflow: hidden;
   }
 
+  figure.pattern-only {
+      border: none !important;
+      height: 100% !important;
+  }
+
```
Comment (kaalleen) [inline print/resources/style.css:584]: `height: 100%` doesn't do anything, but we know the page size and can go for mm values. We good choose e.g. `265mm` here. For both values we do not need `!important`

File: print/templates/pattern_only.html [hunk 5]
```diff
@@ -0,0 +1,25 @@
+    <main class="page pattern-onlyview" id="pattern-display">
```
Comment (kaalleen) [inline print/templates/pattern_only.html:1]: simply write `<main>` without classes and id, especially not with the page class.

File: print/templates/pattern_only.html [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+      <figure class="inksimulation pattern-only operator"
+              data-field-name="pattern-onlyview"
```
Comment (kaalleen) [inline print/templates/pattern_only.html:3]: Change the data-fiel-name to `pattern-onlyview-transform`
This will store the transform value if you reopen the print pdf file again

File: print/templates/pattern_only.html [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+                  <button class="svg-apply">{{ _('Apply to all') }}</button>
+                  <button class="svg-realistic">
+                    {% if loop_index is defined %}
+                    <input type="checkbox" id="realistic-color-block-{{ loop_index }}" data-field-name="block{{ loop_index }}" class="realistic" />
+                    <label for="realistic-color-block-{{ loop_index }}" class="realistic">{{ _('Realistic') }}</label>
+                    {% else %}
+                    <input type="checkbox" id="{{ realistic_id }}" data-field-name="overview" class="realistic" />
+                    <label for="{{ realistic_id }}" class="realistic">{{ _('Realistic') }}</label>
+                    {% endif %}
+                  </button>
```
Comment (kaalleen) [inline print/templates/pattern_only.html:22]: Remove the div with the buttons. We have a template for this (see above)

File: print/templates/pattern_only.html [hunk 8]
```diff
@@ -0,0 +1,25 @@
+    <main class="page pattern-onlyview" id="pattern-display">
+      <figure class="inksimulation pattern-only operator"
+              data-field-name="pattern-onlyview"
+              title="{{ _('Ctrl + Scroll to Zoom') }}"
+              style="cursor: auto;">
+                {{ svg_overview|replace("<li>", "")|replace("</li>", "")|safe }}
+                {% with %}
+                {% set realistic_id='realistic-operator-overview' %}
```
Comment (kaalleen) [inline print/templates/pattern_only.html:8]: We need a specific id here which should not be repeated. Also set the sizing buttons with a template right there.

```
                {% set realistic_id='realistic-pattern-onlyview' %}
                {% include 'ui_svg_action_buttons.html' with context %}
```

File: print/templates/ui.html [hunk 9]
```diff
@@ -54,6 +54,10 @@ <h1>{{ _('Settings') }}</h1>
             <input type="checkbox" class="view" id="operator-detailedview" data-field-name="operator-detailedview" CHECKED />
             <label for="operator-detailedview">{{ _('Operator Detailed View') }}</label>
           </p>
+          <p>
+            <input type="checkbox" class="view" id="pattern-onlyview" data-field-name="pattern-only-view"  />
```
Comment (kaalleen) [inline print/templates/ui.html:58]: Correct the data-fiel-name to `pattern-onlyview`

**Review by kaalleen — CHANGES_REQUESTED** [review]
Ok, I made a more in depth review now.
Hope this is helping.

We could also think about the name. I think "full page pattern" sounds a little nicer than "pattern only". But I'm bad at giving names to new options :D

File: print/templates/pattern_only.html (hunk 7 above)
Comment (cvick32) [reply print/templates/pattern_only.html:22]: The reason I did this was to remove the figcaption that is also included in the ui_svg_action_buttons because I don't think the scale should be printed out if you're just trying to view the pattern. 

Any other way around this?

**Comment by cvick32** [comment]
Thank you so much for your comments. 

Made all requested changes except replacing the svg buttons in pattern_only.html with the template. Please see my comment for the rationale.

File: print/templates/pattern_only.html (hunk 7 above)
Comment (kaalleen) [reply print/templates/pattern_only.html:22]: You could display it at first (could be still interesting while setting it up), but exclude it from the output with css.

In style.css in line 1179 (@media print) you can add

```

    div.pattern-onlyview figcaption {
        display: none;
    }
```

**Comment by cvick32** [comment]
Put the scale in the initial view and exclude it when printing. 

Let me know if there's anything else you need from me. 

 Thanks

**Comment by kaalleen** [comment]
Is looking good so far.
We just need to remove the two empty lines in print.py - otherwise our style checking system will fail on this.

**Comment by cvick32** [comment]
Cool, just removed those lines. Let me know if there's anything else!

**Comment by kaalleen** [comment]
Looks good to me. Thank you for your work!!

**Review by kaalleen — APPROVED** [review]

**Comment by lexelby** [comment]
Awesome!  I love this!  Thanks for all of your work on this.  One very minor request: could you add "Page ##/##" on the new page like all the rest?

**Comment by lexelby** [comment]
@cvick32 I'm going to add you as a collaborator on this project.  That will let you push a branch to this repo, something like `cvick32/pdf-pattern-only`.  Once you do that, our CI will automatically build a test version of your branch including installers for all platforms.  You can then close this PR and make a new PR from that branch, and the CI status will show up in the new PR.

**Comment by kaalleen** [comment]
I think @cvick32 really wanted a blank page with only the pattern on it. Maybe we can make the page numbers an option too?

**Comment by lexelby** [comment]
Oh I thought the page numbers were set to only appear in the preview and not show up on the printed page.  I'm not at my computer so I can't check whether that's true.

**Comment by cvick32** [comment]
Hi again, 

 Sorry I got a little busy with some other stuff, and I think the collaborator invite you sent timed out. Would you mind resending it? 

 Thanks

File: lib/extensions/print_pdf.py [hunk 10]
```diff
@@ -274,6 +274,8 @@ def render_html(self, stitch_plan, overview_svg, selected_palette):
                 'client_detailedview': False,
                 'operator_overview': True,
                 'operator_detailedview': True,
+                'pattern_onlyview': False,
+                'pattern_onlyview': False,
```
Comment (poisa) [inline lib/extensions/print_pdf.py:278]: This looks like it got added twice. (Sorry for the drive-by comment)

Comment (cvick32) [reply lib/extensions/print_pdf.py]: Thanks for catching that!

**Comment by cvick32** [comment]
Hey @lexelby, 

  I just confirmed that the page number does appear in the output PDF when it's exported. Would you still like me to include it on the pattern-only page?

 Cole

**Comment by kaalleen** [comment]
I've re-send te invitation.

What do you think about the page number? Would it be disturbing for what you are planing to do with the pattern only page?

**Comment by cvick32** [comment]
Thanks for resending the invitation. Should I close this PR and push a new branch to this repo?

The page numbers would get in the way of my particular use case, but maybe we can think of a workaround if it's necessary. Maybe a "Toggle Page Numbers" checkbox or something.

**Comment by kaalleen** [comment]
> Thanks for resending the invitation. Should I close this PR and push a new branch to this repo?

Yes, that would be great. Don't worry if the linux build will output an error. It's not your fault, it just seems to have a problem at the moment. We can pull in a fix later (it's a dependency issue).

I really don't care about the page number, but it seems to be important for Lex. I suppose he is very busy at the moment. So let's go for the display page number checkbox?

Since we are on it. I still think that we could find a better name for "Pattern Only View". My suggestion above was "Full Page Pattern". What do you think?

**Comment by lexelby** [comment]
Sorry!  Things have been very busy over here.

Ideally we should show the page number in the preview, and tag it with a CSS class so that it doesn't appear on the printout.  That's how we did the other page numbers.  If you can't figure that out, then it's fine to skip the page number.
